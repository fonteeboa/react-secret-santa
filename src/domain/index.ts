export interface Participant {
    name: string;
}

export interface PlayerFormProps {
    onAddParticipant: (participant: Participant) => void;
}

export interface PlayerListProps {
    participants: Participant[];
    onRemoveParticipant: (index: number) => void;
}

export interface ResultListProps {
    assignments: Map<Participant, Participant>;
}
