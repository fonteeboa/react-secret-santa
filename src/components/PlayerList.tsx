import React from 'react';
import { List, Button } from 'antd';
import VirtualList from 'rc-virtual-list';
import { Participant, PlayerListProps } from '../domain';
import { useTranslation } from 'react-i18next';

const PlayerList: React.FC<PlayerListProps> = ({ participants, onRemoveParticipant }) => {
  const { t } = useTranslation();
  const ContainerHeight = 400;

  return (
    <List bordered className='playerList'>
      <VirtualList
        data={participants}
        height={ContainerHeight}
        itemHeight={47}
        itemKey={(participant) => participant.name}
      >
        {(participant: Participant, index: number) => (
          <List.Item
            key={participant.name}
            actions={[
              <Button type="link" danger onClick={() => onRemoveParticipant(index)}>
                {t('remove')}
              </Button>
            ]}
          >
            {participant.name}
          </List.Item>
        )}
      </VirtualList>
    </List>
  );
};

export default PlayerList;
