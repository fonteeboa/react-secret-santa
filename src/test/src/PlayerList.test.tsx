import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import PlayerList from '../../components/PlayerList';
import { Participant } from '../../domain';
import { useTranslation } from 'react-i18next';
import '../mocks/matchMediaMock';

// Mock do hook useTranslation
jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: string) => {
      if (key === 'remove') {
        return 'Remover';
      }
      return key;
    },
  }),
}));

describe('PlayerList Component', () => {
  const participant1: Participant = { name: 'John' };
  const participant2: Participant = { name: 'Jane' };

  test('renders the player list correctly', () => {
    render(<PlayerList participants={[participant1, participant2]} onRemoveParticipant={jest.fn()} />);

    expect(screen.getByText('John')).toBeInTheDocument();
    expect(screen.getByText('Jane')).toBeInTheDocument();
    expect(screen.getAllByText('Remover')).toHaveLength(2);
  });

  test('calls onRemoveParticipant when remove button is clicked', () => {
    const onRemoveParticipant = jest.fn();
    render(<PlayerList participants={[participant1, participant2]} onRemoveParticipant={onRemoveParticipant} />);

    fireEvent.click(screen.getAllByText('Remover')[0]);

    expect(onRemoveParticipant).toHaveBeenCalledWith(0);
  });

  test('renders an empty player list correctly', () => {
    render(<PlayerList participants={[]} onRemoveParticipant={jest.fn()} />);

    expect(screen.queryByText('John')).not.toBeInTheDocument();
    expect(screen.queryByText('Jane')).not.toBeInTheDocument();
    expect(screen.queryByText('Remover')).not.toBeInTheDocument();
  });

  test('matches the snapshot', () => {
    const { asFragment } = render(<PlayerList participants={[participant1, participant2]} onRemoveParticipant={jest.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
