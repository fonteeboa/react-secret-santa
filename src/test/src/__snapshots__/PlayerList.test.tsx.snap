// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayerList Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ant-list ant-list-split ant-list-bordered playerList css-dev-only-do-not-override-1uq9j6g"
  >
    <div
      class="ant-spin-nested-loading css-dev-only-do-not-override-1uq9j6g"
    >
      <div
        class="ant-spin-container"
      >
        <div
          class="rc-virtual-list"
          style="position: relative;"
        >
          <div
            class="rc-virtual-list-holder"
            style="height: 400px; overflow-y: hidden;"
          >
            <div>
              <div
                class="rc-virtual-list-holder-inner"
                style="display: flex; flex-direction: column;"
              >
                <li
                  class="ant-list-item"
                >
                  John
                  <ul
                    class="ant-list-item-action"
                  >
                    <li>
                      <button
                        class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-link ant-btn-dangerous"
                        type="button"
                      >
                        <span>
                          Remover
                        </span>
                      </button>
                    </li>
                  </ul>
                </li>
                <li
                  class="ant-list-item"
                >
                  Jane
                  <ul
                    class="ant-list-item-action"
                  >
                    <li>
                      <button
                        class="ant-btn css-dev-only-do-not-override-1uq9j6g ant-btn-link ant-btn-dangerous"
                        type="button"
                      >
                        <span>
                          Remover
                        </span>
                      </button>
                    </li>
                  </ul>
                </li>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
